### PR TITLE
sql: enable dependency tracking and support for UDFs calling other UDFs

### LINKED ISSUE #119930: sql/schemachanger: add support for tracking dependencies when UDFs call other UDFs
Today the ability for UDFs to invoke other UDFs is disabled because the schema changer does not have the ability to track dependencies and block DROP operations correctly. To help support this functionality we need to update CREATE FUNCTION to track these dependencies and have DROP FUNCTION use this information to prevent drop operations.

Later on the same information should be usable for supporting cascaded drops.

EPIC: CRDB-19398

Jira issue: CRDB-36396

Epic CRDB-19398

### LINKED ISSUE #120353: sql: drop schema cascade can cause cross schema function references to break
Currently, DROP SCHEMA CASCADE in the legacy schema changer will clean up functions under the schema without properly fixing back references in the originating tables. So, if a default column expression exists references this function, that will be left and only the internal references will be cleaned up. This can lead to corrupt descriptor states where writes / reads may not be possible due to validation errors.

For example:

```
CREATE SCHEMA altSchema;
CREATE FUNCTION altSchema.f() RETURNS INT LANGUAGE SQL AS $$ SELECT 1 $$;
CREATE TABLE t(j int default altSchema.f() CHECK (altSchema.f() > 0));

BEGIN;
drop schema sc1 cascade;
COMMIT;

-- trouble
SELECT * FROM t;
```

Jira issue: CRDB-36645

### PR DESCRIPTION
**Pull request by fqazi**
Previously, we did not support UDFs invoking other UDFs because the dependency tracking infrastructure was missing. That meant it was possible for a UDF to be broken if a dependent was dropped. To address this this patch does the following:

- Inside optbuilder enable support for this feature (Note: This doesn't mean its production ready but to allow us to test things out).
- Add dependency tracking support for invocation to other UDFs
- Use the dependency tracing inside the legacy / declarative CREATE FUNCTION so that the descriptor reflects this information
- If any dependent object exists block DROP FUNCTION operations

This PR does not add DROP FUNCTION CASCADE yet, which will need a bit more work since it needs to deal with all sorts of other dependencies.

Informs: #88198
Fixes: #120353
Fixes: https://github.com/cockroachdb/cockroach/issues/119930

### CODE REVIEW

**Review by DrewKimball — COMMENTED** [review]
This is exciting work! I have mostly nits and test requests, apart from a bug in the optbuilder code.

> Reviewed 4 of 4 files at r1, 21 of 21 files at r2, 29 of 29 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-:-NsMYNKkBCnj_sd5vuep:b-3egll8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @mgartner, and @yuzefovich)

___
*[`pkg/sql/create_function.go` line 455 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMN5hU-XT2XSwqMlW-:-NsMN5hU-XT2XSwqMlW0:bvzkstn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/create_function.go#L455)):*
> ```Go
> 	// Add forward references to UDF descriptor.
> 	udfDesc.DependsOn = backrefTblIDs.Ordered()
> 	//udfDesc.DependsOnFunction =
> ```

[nit] Left over line here.
___
*[`pkg/sql/catalog/descpb/structured.proto` line 1657 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMPERv2tGxIJykNGKK:-NsMPERv2tGxIJykNGKL:b6tp3cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/descpb/structured.proto#L1657)):*
> ```proto
>   repeated uint32 depends_on_types = 14 [(gogoproto.casttype) = "ID"];
> ```

[nit] extra newline
___
*[`pkg/sql/catalog/descpb/structured.proto` line 1674 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMPGcM0B316bZIlbl2:-NsMPGcM0B316bZIlbl3:b-nob3d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/descpb/structured.proto#L1674)):*
> ```proto
> 
>   // DependsOnFunctions are the user defined functions that this function
>   // qdepends on.
> ```

[nit] `qdepends` -> `depends`
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMPbB9EBx0aufUlNW3:-NsMPbB9EBx0aufUlNW4:becpivh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/funcdesc/func_desc.go#L292)):*
> ```Go
> 
> 	// Currently, we don't support cross function references yet.
> 	// So here we assume that all inbound references are from tables.
> ```

[nit] This comment is outdated.
___
*[`pkg/sql/opt/optbuilder/routine.go` line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMQlaL9Gb2W00EmJZN:-NsMQlaL9Gb2W00EmJZO:bi89bsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/opt/optbuilder/routine.go#L98)):*
> ```Go
> 
> 	if b.trackSchemaDeps {
> 		b.schemaFunctionDeps.Add(int(o.Oid))
> ```

This will miss a UDF used as a data source like `SELECT * FROM f()`, as well as a stored proc like `CALL p()`. This would probably be the best location: https://github.com/cockroachdb/cockroach/blob/f4dc2b59541e11cf06e2b948c7c37bf1aed21868/pkg/sql/opt/optbuilder/routine.go#L164-L165

Also, we should set `trackSchemaDeps` to false (and then defer resetting it back to true) like we do for views [here](https://github.com/cockroachdb/cockroach/blob/f4dc2b59541e11cf06e2b948c7c37bf1aed21868/pkg/sql/opt/optbuilder/select.go#L360-L363), so that we don't pick up dependencies from within the nested routine's body.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMUVkeEm8DNak-m-uk:-NsMUVkeEm8DNak-m-ul:bqwjrdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8962717b122516720e9fd59f84bcace01ee84ac/pkg/sql/schemachanger/scpb/elements.proto#L727)):*
> ```proto
>   repeated uint32 uses_sequence_ids = 6 [(gogoproto.customname) = "UsesSequenceIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
>   repeated uint32 uses_type_ids = 7 [(gogoproto.customname) = "UsesTypeIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
>   repeated FunctionReference uses_functions = 8  [(gogoproto.nullable) = false];
> ```

[nit] Would it be nicer to do an ID slice instead like for type tracking? Do you anticipate we'll need to track other fields in the future?
___
*[`pkg/sql/logictest/testdata/logic_test/drop_function` line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMTDMQ-bkxlpiJNgKq:-NsMTDMQ-bkxlpiJNgKr:bl2ja87) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/logictest/testdata/logic_test/drop_function#L288)):*
> ```
> 
> statement ok
> DROP FUNCTION f_called_by_b2;
> ```

Can we have similar tests for a depended-on function where we:
* try to CREATE OR REPLACE it
* try to rename it
* try to change its schema

Here are some tests I wrote a while ago while thinking about this issue: https://gist.github.com/DrewKimball/47390e110fc3dc3fae11cc96adcd9579

**Review by yuzefovich — COMMENTED** [review]
Nice work! I only have some nits and will defer to Drew for approval.

> Reviewed 4 of 4 files at r1, 21 of 21 files at r2, 29 of 29 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-:-Nsjlbbx5UljBD6_AYw5:bsvbep5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @fqazi, and @mgartner)

___
*[`pkg/sql/create_function.go` line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-Nsjc_8aAj93RzWNVg5s:-Nsjc_8aAj93RzWNVg5t:b-i12lw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/create_function.go#L77)):*
> ```Go
> 		}
> 		if dbID := funcDesc.GetParentID(); dbID != n.dbDesc.GetID() && dbID != keys.SystemDatabaseID {
> 			return pgerror.Newf(pgcode.FeatureNotSupported, "the function cannot refer to other databases")
> ```

nit: should we make this error message more user-friendly? Something like `"dependent function %s cannot refer to other databases", funcDesc.Name()`.
___
*[`pkg/sql/create_function.go` line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsjhFr3793fztoqg2xG:-NsjhFr3793fztoqg2xH:b-p9sl87) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/create_function.go#L180)):*
> ```Go
> 
> func (n *createFunctionNode) replaceFunction(udfDesc *funcdesc.Mutable, params runParams) error {
> 	// TODO(chengxiong): add validation that the function is not referenced. This
> ```

nit: this TODO should probably be removed.
___
*[`pkg/sql/opt_exec_factory.go` line 1871 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-Nsjhrbk2UHkJXCvzIAS:-Nsjhrbk2UHkJXCvzIAT:b-s81t9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/opt_exec_factory.go#L1871)):*
> ```Go
> 	}
> 
> 	planDeps, typeDepSet, _, err := toPlanDependencies(deps, typeDeps, intsets.Fast{})
> ```

nit: add inlined comment `/* funcDeps */`.
___
*[`pkg/sql/catalog/validate.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsjjsFT4hFQTr7pxRXh:-NsjjsFT4hFQTr7pxRXi:b-561g16) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/validate.go#L125)):*
> ```Go
> 
> // ValidateOutboundFunctionRef validates outbound reference to a function descriptor
> // depID from descriptor selfID.
> ```

nit: I don't see a mention of `selfID` in the argument list nor in the code (ditto for the comment on `ValidateOutboundTableRef`).
___
*[`pkg/sql/catalog/descpb/structured.proto` line 1676 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsjiRu86YQvuM_5t-kp:-NsjiRu86YQvuM_5t-kq:b-5pvepm) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/descpb/structured.proto#L1676)):*
> ```proto
>   // qdepends on.
>   repeated uint32 depends_on_functions = 22  [(gogoproto.casttype) = "ID"];
>   // Next field id is 22
> ```

nit: now it's 23
___
*[`pkg/sql/logictest/testdata/logic_test/drop_function` line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMTDMQ-bkxlpiJNgKq:-NsjkT5U996YRChuocky:b68a0tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/logictest/testdata/logic_test/drop_function#L288)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can we have similar tests for a depended-on function where we:
* try to CREATE OR REPLACE it
* try to rename it
* try to change its schema

Here are some tests I wrote a while ago while thinking about this issue: https://gist.github.com/DrewKimball/47390e110fc3dc3fae11cc96adcd9579
</blockquote></details>

We probably should add some test cases to `udf_rewrite` files too.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-Nsjc3tS0J0WCJxn7kJ_:-Nsjc3tS0J0WCJxn7kJa:b-mwa6u9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ccbcf1673e6562d7c6d6203d015697d7382ab3/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L101)):*
> ```
> 
> statement ok
> CREATE FUNCTION f_using_udf() RETURNS INT LANGUAGE SQL AS $$ SELECT [FUNCTION $fn_oid]() $$;
> ```

nit: we should also invoke this UDF now.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_unsupported` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsjcCrw-20G29snEu4u:-NsjcCrw-20G29snEu4v:b-kms4de) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ccbcf1673e6562d7c6d6203d015697d7382ab3/pkg/sql/logictest/testdata/logic_test/udf_unsupported#L39)):*
> ```
> 
> 
> subtest disallow_udf_in_views_and_udf
> ```

nit: the subtest name should be adjusted. Also probably some of the test cases should be moved into a different file (since they are no longer "unsupported").

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119932)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @yuzefovich)

___
*[`pkg/sql/create_function.go` line 455 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMN5hU-XT2XSwqMlW-:-NsiXlq4AN295jWk-hlz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/create_function.go#L455)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Left over line here.
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/descpb/structured.proto` line 1657 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMPERv2tGxIJykNGKK:-NsigsRx6DHbPAB6IBsK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/descpb/structured.proto#L1657)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] extra newline
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/descpb/structured.proto` line 1674 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMPGcM0B316bZIlbl2:-NsigtHIADOt4jgcVZlr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/descpb/structured.proto#L1674)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] `qdepends` -> `depends`
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMPbB9EBx0aufUlNW3:-NsiiAUb19WxJm8fb4LP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/catalog/funcdesc/func_desc.go#L292)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] This comment is outdated.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/routine.go` line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMQlaL9Gb2W00EmJZN:-NsikQwf34PeCGpbdL-Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/opt/optbuilder/routine.go#L98)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This will miss a UDF used as a data source like `SELECT * FROM f()`, as well as a stored proc like `CALL p()`. This would probably be the best location: https://github.com/cockroachdb/cockroach/blob/f4dc2b59541e11cf06e2b948c7c37bf1aed21868/pkg/sql/opt/optbuilder/routine.go#L164-L165

Also, we should set `trackSchemaDeps` to false (and then defer resetting it back to true) like we do for views [here](https://github.com/cockroachdb/cockroach/blob/f4dc2b59541e11cf06e2b948c7c37bf1aed21868/pkg/sql/opt/optbuilder/select.go#L360-L363), so that we don't pick up dependencies from within the nested routine's body.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMUVkeEm8DNak-m-uk:-NsixkAG52hINr7MttIm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8962717b122516720e9fd59f84bcace01ee84ac/pkg/sql/schemachanger/scpb/elements.proto#L727)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Would it be nicer to do an ID slice instead like for type tracking? Do you anticipate we'll need to track other fields in the future?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/drop_function` line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsMTDMQ-bkxlpiJNgKq:-NsnYnm5CbYZehsxvuT3:b-4q0tbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec11fa5902c38d326e73ee174dd30eef5e4a611/pkg/sql/logictest/testdata/logic_test/drop_function#L288)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We probably should add some test cases to `udf_rewrite` files too.
</blockquote></details>

Good point, I added tests to confirm this is broken. I have opened:  #120351 for fixing it.

**Review by DrewKimball — APPROVED** [review]
:lgtm: Great work!

> Reviewed 14 of 63 files at r4, 55 of 55 files at r8, 2 of 2 files at r9, 24 of 24 files at r10, 30 of 30 files at r11, 4 of 4 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-:-NsoIHENAtqx7oBxGYPT:bolbkdq)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi, @mgartner, and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/builder.go` line 147 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsoDa8xEQu4yXtWShXf:-NsoDa8xEQu4yXtWShXg:b-qqcdwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7fe9bb278d834c05efa1ee1409fa288f75aa5be/pkg/sql/opt/optbuilder/builder.go#L147)):*
> ```Go
> 	// insideUDF is true when the current expressions are being built within a
> 	// UDF.
> 	insideUDF int
> ```

[nit] Comment is outdated.

Also, I think it'd be fine to keep as a bool, and combine restoration of the previous value with `trackSchemaDeps`:
```
oldTrackingSchemaDeps := b.trackSchemaDeps
oldInsideUDF := b.insideUDF
defer func() {
  b.trackSchemaDeps = oldTrackingSchemaDeps
  b.insideUDF = oldInsideUDF
}()
```
Up to you, though.
___
*[`pkg/sql/opt/optbuilder/routine.go` line 264 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/119932#-NsoDvFL1Q5A8JKGyaTh:-NsoDvFL1Q5A8JKGyaTi:b-v0o73) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7fe9bb278d834c05efa1ee1409fa288f75aa5be/pkg/sql/opt/optbuilder/routine.go#L264)):*
> ```Go
> 	// boolean will not be sufficient to track whether or not we are in a UDF.
> 	// We'll need to track the depth of the UDFs we are building expressions
> 	// within.
> ```

[nit] Outdated comment.

**Comment by fqazi** [comment]
@DrewKimball @yuzefovich  TFTR!

bors r+

**Pull request merged**
